dali/operators/image/remap/displacement_filter_impl_gpu.cuh

```diff
@@ -47,155 +59,154 @@ __device__ inline void GetPixelValueMultiC(int h, int w,
+  __device__ __host__ inline void operator() (Displacement &, const void *) {}
 };
 
 template <class Displacement>
 struct PrepareParam<Displacement, true> {
- public:
-  __device__ __host__ inline void operator() (Displacement &displace,  // NOLINT(*)
-                                              const void * raw_params, const int n) {
-    const typename Displacement::Param * const params =
-      reinterpret_cast<const typename Displacement::Param *>(raw_params);
-    displace.param = params[n];
+  __device__ __host__ inline void operator()(Displacement &displace, const void *raw_params) {
+    const auto *const params = reinterpret_cast<const typename Displacement::Param *>(raw_params);
+    displace.param = *params;
   }
 };
 
 template <typename T, bool per_channel_transform, class Displacement, DALIInterpType interp_type>
-__global__
-void DisplacementKernel(const T *in, T* out,
-                        const int N, const Index * shapes, const bool has_mask,
-                        const int * mask, const void * raw_params, const Index pitch,
-                        const T fill_value,
-                        Displacement displace) {
-  // block per image
-  for (int n = blockIdx.x; n < N; n += gridDim.x) {
-    const int H = shapes[n * pitch + 0];
-    const int W = shapes[n * pitch + 1];
-    const int C = shapes[n * pitch + 2];
-    const Index offset = shapes[n * pitch + 3];
-    const bool m = !has_mask || mask[n];
-    PrepareParam<Displacement> pp;
-    pp(displace, raw_params, n);
-    // thread per pixel
-    const T *image_in = in + offset;
-    T *image_out = out + offset;
-    for (int out_idx = threadIdx.x; out_idx < H * W * C; out_idx += blockDim.x) {
-      if (m) {
-        const int c = out_idx % C;
-        const int w = (out_idx / C) % W;
-        const int h = (out_idx / W / C);
-
-        image_out[out_idx] = GetPixelValueSingleC<T, Displacement, interp_type>(h, w, c, H, W, C,
-            image_in, displace, fill_value);
-      } else {
-        image_out[out_idx] = image_in[out_idx];
-      }
+__global__ void DisplacementKernel(const DisplacementSampleDesc *samples,
+                                   const kernels::BlockDesc<1> *blocks, const T fill_value,
+                                   Displacement displace) {
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
+
+  auto *image_out = reinterpret_cast<T *>(sample.output);
+  const auto *image_in = reinterpret_cast<const T *>(sample.input);
+
+  const int H = sample.shape[0];
+  const int W = sample.shape[1];
+  const int C = sample.shape[2];
+  const bool m = sample.mask;
+
+  PrepareParam<Displacement> pp;
+  pp(displace, sample.raw_params);
+
+  auto start = block.start.x;
+  auto end = block.end.x;
+  for (int64_t out_idx = threadIdx.x + start; out_idx < end; out_idx += blockDim.x) {
+    if (m) {
+      const int c = out_idx % C;
+      const int w = (out_idx / C) % W;
+      const int h = (out_idx / W / C);
```

Comment:
```suggestion
      int64_t idx = out_idx;
      const int c = idx % C;
      idx /= C;
      const int w = idx % W;
      idx /= W;
      const int h = idx;
```
or at least
```suggestion
      const int c = out_idx % C;
      const int w = (out_idx / C) % W;
      const int h = (out_idx / C / W);
```
The way it was written, it prevented optimization of the last two divisions.

Reply:
done